PR: https://github.com/odoo/odoo/pull/264299

From: ef13d68b3c895d973f36c2220eecdb81ccafcf7c
From: jodh-odoo
Date: 2026-06-23 07:27:13

Breaking data model changes scores: del:1 + add:0, change matches:
-    group_mrp_reception_report = fields.Boolean("Allocation Report for Manufacturing Orders", implied_group='mrp.group_mrp_reception_report')

Total Changes: 80

[IMP] mrp, stock: allocation report setting by picking type

Allocation report settings
==========================
Before this commit, there were two settings for allocation reports:
* Reception Report (`group_stock_reception_report`);
* Allocation Report for Manufacturing Orders
  (`group_mrp_reception_report`).

Now, this setting is moved on the operation types and can be
enabled/disabled by operation type (except for out operations, it's
always disabled in this case) and this commit adapts code to this
change.

Additional change
=================
This commit moves the `stock.picking` `_get_autoprint_report_actions`
code part which fetches moves' allocation report action to create
`stock.move` own action: `action_open_allocation_report`.
The purpose is to be able to use this part from different places (from
the Barcode application for example) without duplicate the code.

task-4894566

Part-of: odoo/odoo#264299
Related: odoo/enterprise#117220
Related: odoo/upgrade#10320
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp/models/mrp_production.py
+++ b/addons/mrp/models/mrp_production.py
@@ -760,25 +760,26 @@ class MrpProduction(models.Model):
     @api.depends('state', 'move_finished_ids')
     def _compute_show_allocation(self):
         self.show_allocation = False
-        if not self.env.user.has_group('mrp.group_mrp_reception_report'):
-            return
         for mo in self:
-            if not mo.picking_type_id:
-                return
-            lines = mo.move_finished_ids.filtered(lambda m: m.product_id.is_storable and m.state != 'cancel')
-            if lines:
+            if not mo.picking_type_id.auto_show_reception_report:
+                continue
+            moves = mo.move_finished_ids.filtered(lambda m: m.product_id.is_storable and m.state != 'cancel')
+            if moves:
                 allowed_states = ['confirmed', 'partially_available', 'waiting']
                 if mo.state == 'done':
                     allowed_states += ['assigned']
-                wh_location_ids = self.env['stock.location']._search([('id', 'child_of', mo.picking_type_id.warehouse_id.view_location_id.id), ('usage', '!=', 'supplier')])
+                wh_location_ids = self.env['stock.location']._search([
+                    ('id', 'child_of', mo.picking_type_id.warehouse_id.view_location_id.id),
+                    ('usage', '!=', 'supplier'),
+                ])
                 if self.env['stock.move'].search_count([
                     ('state', 'in', allowed_states),
                     ('product_qty', '>', 0),
                     ('location_id', 'in', wh_location_ids),
                     ('raw_material_production_id', 'not in', mo.ids),
-                    ('product_id', 'in', lines.product_id.ids),
+                    ('product_id', 'in', moves.product_id.ids),
                     '|', ('move_orig_ids', '=', False),
-                        ('move_orig_ids', 'in', lines.ids)], limit=1):
+                        ('move_orig_ids', 'in', moves.ids)], limit=1):
                     mo.show_allocation = True
 
     @api.depends('product_uom_qty', 'date_start')
@@ -2376,12 +2377,14 @@ class MrpProduction(models.Model):
                     'res_id': self.id,
                     'target': 'main',
                 }
-            elif self.env.user.has_group('mrp.group_mrp_reception_report'):
+            else:
                 mos_to_show = self.filtered(lambda mo: mo.picking_type_id.auto_show_reception_report)
-                lines = mos_to_show.move_finished_ids.filtered(lambda m: m.product_id.is_storable and m.state != 'cancel' and m.picked and not m.move_dest_ids)
-                if lines:
-                    if any(mo.show_allocation for mo in mos_to_show):
-                        another_action = mos_to_show.action_view_reception_report()
+                has_lines_to_assign = any(
+                    (m.product_id.is_storable and m.state != 'cancel' and m.picked and not m.move_dest_ids)
+                    for m in mos_to_show.move_finished_ids
+                )
+                if has_lines_to_assign and any(mo.show_allocation for mo in mos_to_show):
+                    another_action = mos_to_show.action_view_reception_report()
             if report_actions:
                 return {
                     'type': 'ir.actions.client',
@@ -2391,9 +2394,7 @@ class MrpProduction(models.Model):
                         'anotherAction': another_action,
                     }
                 }
-            if another_action:
-                return another_action
-            return True
+            return another_action or True
         context = {
             k: False if k.startswith('skip_') else v
             for k, v in self.env.context.items()
@@ -3070,30 +3071,26 @@ class MrpProduction(models.Model):
                 action = self.env.ref("mrp.label_manufacture_template").report_action(labels_to_print.ids, config=False)
                 clean_action(action, self.env)
                 report_actions.append(action)
-        if self.env.user.has_group('mrp.group_mrp_reception_report'):
-            reception_reports_to_print = self.filtered(
-                lambda p: p.picking_type_id.auto_print_mrp_reception_report
-                          and p.picking_type_id.code == 'mrp_operation'
-                          and p.move_finished_ids.move_dest_ids
-            )
-            if reception_reports_to_print:
-                action = self.env.ref('stock.stock_reception_report_action').report_action(reception_reports_to_print, config=False)
-                action['context'] = dict({'default_production_ids': reception_reports_to_print.ids}, **self.env.context)
-                clean_action(action, self.env)
-                report_actions.append(action)
-            reception_labels_to_print = self.filtered(lambda p: p.picking_type_id.auto_print_mrp_reception_report_labels and p.picking_type_id.code == 'mrp_operation')
-            if reception_labels_to_print:
-                moves_to_print = reception_labels_to_print.move_finished_ids.move_dest_ids
-                if moves_to_print:
-                    # needs to be string to support python + js calls to report
-                    quantities = ','.join(str(qty) for qty in moves_to_print.mapped(lambda m: math.ceil(m.product_uom_qty)))
-                    data = {
-                        'docids': moves_to_print.ids,
-                        'quantity': quantities,
-                    }
-                    action = self.env.ref('stock.label_picking').report_action(moves_to_print, data=data, config=False)
-                    clean_action(action, self.env)
-                    report_actions.append(action)
+
+        # Filters MOs and their dest. moves to print reception report and labels.
+        orders_for_reception_report = self.env['mrp.production']
+        print_label_move_ids = set()
+        for order in self:
+            if not order.picking_type_id.auto_show_reception_report or order.picking_type_id.code != 'mrp_operation' or not order.move_finished_ids.move_dest_ids:
+                continue
+            if order.picking_type_id.auto_print_mrp_reception_report:
+                orders_for_reception_report |= order
+            if order.picking_type_id.auto_print_mrp_reception_report_labels:
+                print_label_move_ids.update(order.move_finished_ids.move_dest_ids.ids)
+        if orders_for_reception_report:
+            action = self.env.ref('stock.stock_reception_report_action').report_action(orders_for_reception_report, config=False)
+            action['context'] = dict(default_production_ids=orders_for_reception_report.ids, **self.env.context)
+            clean_action(action, self.env)
+            report_actions.append(action)
+        if print_label_move_ids:
+            action = self.env['stock.move'].browse(print_label_move_ids).action_print_reception_report()
+            clean_action(action, self.env)
+            report_actions.append(action)
         if self.env.user.has_group('stock.group_production_lot'):
             productions_to_print = self.filtered(lambda p: p.picking_type_id.auto_print_done_mrp_lot and p.move_finished_ids.move_line_ids.lot_id)
             productions_by_print_formats = productions_to_print.grouped(lambda p: p.picking_type_id.done_mrp_lot_label_to_print)

--- a/addons/mrp/models/res_config_settings.py
+++ b/addons/mrp/models/res_config_settings.py
@@ -17,7 +17,6 @@ class ResConfigSettings(models.TransientModel):
     group_mrp_routings = fields.Boolean("MRP Work Orders",
         implied_group='mrp.group_mrp_routings')
     group_unlocked_by_default = fields.Boolean("Unlock Manufacturing Orders", implied_group='mrp.group_unlocked_by_default')
-    group_mrp_reception_report = fields.Boolean("Allocation Report for Manufacturing Orders", implied_group='mrp.group_mrp_reception_report')
 
     def set_values(self):
         routing_before = self.env.user.has_group('mrp.group_mrp_routings')
